src/datastore/pg-write-store.ts

```diff
@@ -1696,21 +1692,82 @@ export class PgWriteStore extends PgStore {
         SELECT tx_id FROM inserted
       `;
       txIds.push(...result.map(r => r.tx_id));
+      // The incoming mempool tx might have already been settled
+      // We need to mark them as pruned to avoid inconsistent tx state
+      const pruned_tx = await sql<{ tx_id: string }[]>`
```

Comment:
Instead of multiple sql queries after the inserting into `mempool_txs`, can you instead modify the above query so that if a tx is already settled, then it should be inserted with `pruned=true`?

Reply:
The added query is run using the index: `Index Scan using txs_tx_id_index on txs  (cost=0.00..12.04 rows=1 width=33)`; so mempool transactions event should be processed within 20ms.

Reply:
Is there a reason it can't be done in the first query with a CTE or similar?

Reply:
Not sure how to write the CTE query to insert `${sql(values)}` and override the pruned state, and in the meantime update the tx already exists with correct pruned state, and also get the correct number of new mempool tx count. Looks like it will be more complex and maybe less performant.

Can you give me an example so I can follow?

Reply:
It's likely more performant to perform the updates in one query as it gives the postgres query planner more room to optimize. I can take a look at helping with the PR later unless we can determine that this is high priority right now

Reply:
Is this what you're suggesting?

```typescript
const result = await sql<{ tx_id: string }[]>`
  WITH inserted AS (
    INSERT INTO mempool_txs ${sql(values)}
    ON CONFLICT ON CONSTRAINT unique_tx_id DO NOTHING
    RETURNING tx_id
  ),
  settled AS (
    SELECT tx_id
    FROM txs
    WHERE
      tx_id IN ${sql(values.map(b => b.tx_id))} AND
      canonical = true AND
      microblock_canonical = true AND
      status IN ${sql([
        DbTxStatus.Success,
        DbTxStatus.AbortByResponse,
        DbTxStatus.AbortByPostCondition,
      ])}
  ),
  pruned AS (
    UPDATE mempool_txs
    SET pruned = true
    WHERE
      tx_id IN (SELECT tx_id FROM settled) AND
      pruned = false
    RETURNING tx_id
  ),
  count_update AS (
    UPDATE chain_tip SET
      mempool_tx_count = mempool_tx_count + (SELECT COUNT(*) FROM inserted) - (SELECT COUNT(*) FROM pruned),
      mempool_updated_at = NOW()
  )
  SELECT tx_id FROM inserted
`;
```

Reply:
That looks about right. Is the `status IN ...` necessary? And would it be possible to avoid passing in tx_id twice? E.g.
```sql
...
settled AS (
  SELECT tx_id
  FROM txs
  WHERE
    tx_id IN (SELECT tx_id FROM inserted) AND
    canonical = true AND
    microblock_canonical = true
)
...
```

Reply:
> Is the status IN ... necessary?

I don't think so, but it's in the `reconcileMempoolStatus` function. I thought there were other possibilities.

The `inserted` is a subset of the original transactions, because duplicated tx_id will be excluded.

Reply:
The cost is way higher than checking current inserted transactions.

![image](https://github.com/hirosystems/stacks-blockchain-api/assets/432106/4717c72f-9967-47c8-b4c1-471f3f6c6c3e)

Reply:
Pushed the changes.

Reply:
LGTM, @rafaelcr can you also review this query?